compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,383 @@
+}
+
+// Compute the number of global loads per thread based on the target's
+// max load instruction size.
+static int64_t getNumOfGlobalLoadsPerThread(MemRefType destType,
+                                            int64_t subgroupSize,
+                                            int64_t maxLoadInstructionBits) {
+  // First, compute total number of bytes to load.
+  int64_t elementSize = destType.getElementTypeBitWidth();
+  int64_t totalBits = destType.getNumElements() * elementSize;
+
+  // Then, compute number of bytes each thread should load.
+  assert(totalBits % subgroupSize == 0 &&
+         "Total bits must be divisible by subgroup size");
+  int64_t bitsPerThread = totalBits / subgroupSize;
+
+  int64_t copiesPerThread = bitsPerThread / maxLoadInstructionBits;
+  return copiesPerThread;
+}
+
+// Lowers iree_gpu.coalesced_gather_dma operations within scf.forall loops
+// with thread mapping to amdgpu.gather_to_lds operations inside scf.for loops.
+//
+// In short, looks for patterns like this:
+//   scf.forall (...) in (32, 1) {
+//     %1 = iree_gpu.coalesced_gather_dma %indices, %source into %dest
+//   } {mapping = [#gpu.thread<linear_dim_1>, #gpu.thread<linear_dim_0>]}
+//
+// And replaces with:
+//   %lane_id = gpu.lane_id
+//   scf.for %iv = %c0 to %num_copies step %c1 {
+//     %thread_index = affine.delinearize_index(%subgroup_id, ...)
+//     %store_index = affine.delinearize_index(%lane_id, ...)
+//     amdgpu.gather_to_lds %source[%thread_index] into %dest[%store_index]
+//   }
+struct LowerCoalescedGatherDMAPattern : public OpRewritePattern<scf::ForallOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LowerCoalescedGatherDMAPattern(MLIRContext *context,
+                                 ArrayRef<int64_t> workgroupSize,
+                                 int64_t subgroupSize,
+                                 int64_t maxLoadInstructionBits)
+      : OpRewritePattern<scf::ForallOp>(context), workgroupSize(workgroupSize),
+        subgroupSize(subgroupSize),
+        maxLoadInstructionBits(maxLoadInstructionBits) {}
+
+  LogicalResult matchAndRewrite(scf::ForallOp forallOp,
+                                PatternRewriter &rewriter) const override {
+    // Verify that the forall has the required GPU thread mapping.
+    if (failed(verifyThreadMapping(forallOp))) {
+      return failure();
+    }
+
+    // Ensure this is not nested inside another thread-mapped forall
+    if (forallOp->getParentOfType<scf::ForallOp>()) {
+      auto parentForall = forallOp->getParentOfType<scf::ForallOp>();
+      if (parentForall.getMapping()) {
+        auto parentMapping = *parentForall.getMapping();
+        if (!parentMapping.empty() &&
+            isa<gpu::GPUThreadMappingAttr>(parentMapping[0])) {
+          return rewriter.notifyMatchFailure(
+              forallOp, "nested thread-mapped forall not supported");
+        }
+      }
+    }
+
+    // Check that the forall body contains exactly one DMA op (plus terminator)
```

Comment:
Why do we care if the dma op has a forall op for the parent? It's lowering should be independent of where it is.